### PR TITLE
[webui] Introduces repositories controller

### PR DESCRIPTION
**Pull request by ChrisBr**
- Moved the following actions from project controller to repositories controller
  - #repositories -> #index
  - #add_repositories -> #new
  - #add_repository_from_default_list -> #distributions
  - #save_repository -> #create
  - #update_target -> #update
  - #remove_target -> #destroy
  - #repository_state -> #state
  - #create_dod_repository -> #create_dod_repository
  - #create_flag -> #create_flag
  - #toggle_flag -> #toggle_flag
  - #remove_flag -> #remove_flag
- Moved the following actions from package controller to repositories controller
  - #repositories -> #index
- Created new action in repositories controller
  - create_image_repository
- Removed has_flags mixin
- Introduced acts_as_list to RepositoryArchitecture and PathElement
- Moved tests from project_controller_spec to repositories_controller_spec

#1889

### CODE REVIEW

File: src/api/app/controllers/webui/project_controller.rb [hunk 1]
```diff
@@ -616,9 +428,10 @@ def move_path
 
     if @project.valid?
       @project.store
-      redirect_to({ :action => :repositories, :project => @project}, notice: "Path moved #{params[:direction]} successfully")
+      redirect_to({ :action => :index, controller: :repositories, :project => @project}, notice: "Path moved #{params[:direction]} successfully")
```
Comment (mdeniz) [inline src/api/app/controllers/webui/project_controller.rb]: Can you use the new syntax for the whole hash here?
```ruby
action: :index, controller: :repositories, project: @project
```

File: src/api/app/models/repository.rb [hunk 2]
```diff
@@ -31,6 +31,15 @@ class Repository < ActiveRecord::Base
 
   validates :db_project_id, presence: true
 
+  validate do |repository|
+    repository.path_elements.each do |path_element|
+      next if path_element.valid?
```
Comment (mdeniz) [inline src/api/app/models/repository.rb]: Instead using `next` we can use a reject loop:
```ruby
repository.path_elements.reject(&:valid?).each do |path_element|
  path_element.errors.full_messages.each do |msg|
      errors[:base] << "Path Element: #{msg}"
  end
end
```

**Comment by mdeniz** [comment]
Beside those nitpicks it LGTM, I guess that someone more should review this because is a pretty complex change. @bgeuken, @hennevogel do you mind??

File: src/api/app/controllers/webui/repositories_controller.rb [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+    end
+  end
+
+  # POST project/update_target/:project
+  def update
+    authorize @project, :update?
+    repo = @project.repositories.where(name: params[:repo]).first
+    archs = []
+    archs = params[:arch].keys.map { |arch| Architecture.find_by_name(arch) } if params[:arch]
+    repo.architectures = archs
```
Comment (hennevogel) [inline src/api/app/controllers/webui/repositories_controller.rb]: What is repo is nil?

File: src/api/app/controllers/webui/repositories_controller.rb [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+    repo.save
```
Comment (hennevogel) [inline src/api/app/controllers/webui/repositories_controller.rb]: What is this goes wrong?

File: src/api/app/controllers/webui/repositories_controller.rb (hunk 3 above)
Comment (ChrisBr) [inline src/api/app/controllers/webui/repositories_controller.rb]: This is c&p from the project_controller, never touched it!

File: src/api/app/controllers/webui/repositories_controller.rb (hunk 4 above)
Comment (ChrisBr) [inline src/api/app/controllers/webui/repositories_controller.rb]: This is c&p from the project_controller, never touched it!

Comment (hennevogel) [inline src/api/app/controllers/webui/repositories_controller.rb]: Alright, nevermind then :-)

File: src/api/app/controllers/webui/repositories_controller.rb [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+    if @project.valid? && result
+      @project.store
+      respond_to do |format|
+        flash[:success] = "Successfully removed repository '#{repository.name}'"
+        format.html { redirect_to({ action: :index, project: @project }) }
+        format.js
+      end
+    else
+      respond_to do |format|
+        flash[:error] = "Failed to remove repository: #{@project.errors.full_messages.to_sentence}"
```
Comment (hennevogel) [inline src/api/app/controllers/webui/repositories_controller.rb]: What is only result is nil? Then @project won't have errors.

File: src/api/app/controllers/webui/repositories_controller.rb [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+      flash[:error] = "Can not add image repository: #{repository.errors.full_messages.to_sentence}"
+      respond_to do |format|
+        format.html { redirect_to :back }
+        format.js { render 'create' }
+      end
+    end
+  end
+
+  # POST flag/:project(/:package)
+  # Compatibility routes
```
Comment (hennevogel) [inline src/api/app/controllers/webui/repositories_controller.rb]: Why for POST?

File: src/api/app/views/webui/repositories/index.html.haml [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+      = link_to @project, :controller => :repositories, :action => :index, :project => @project
+    = render :partial => 'shared/repositories', :locals => { :obj => @package }
+  - else
+    %p
+      This package is inherited from the project
+      = link_to @package.project, :action => :index, controller: :repositories, :project => @package.project
+      and changing repository flags has no effect.
+- else
+  %h3
+  = @pagetitle
```
Comment (hennevogel) [inline src/api/app/views/webui/repositories/index.html.haml]: Missing indentation

File: src/api/app/controllers/webui/repositories_controller.rb (hunk 5 above)
Comment (ChrisBr) [inline src/api/app/controllers/webui/repositories_controller.rb]: Good point, changed it to:

```
      msg = "Failed to remove repository: #{@project.errors.full_messages.to_sentence}"
      msg << 'Repository not found.' if @project.valid? && !result
```

File: src/api/app/controllers/webui/repositories_controller.rb (hunk 6 above)
Comment (ChrisBr) [inline src/api/app/controllers/webui/repositories_controller.rb]: Removed comment, already removed it in routes.rb

File: src/api/app/models/repository.rb (hunk 2 above)
Comment (ChrisBr) [inline src/api/app/models/repository.rb]: ok, done!

File: src/api/app/views/webui/repositories/index.html.haml (hunk 7 above)
Comment (ChrisBr) [inline src/api/app/views/webui/repositories/index.html.haml]: Done

**Comment by ChrisBr** [comment]
@mdeniz & @hennevogel fixed your suggestions, please review again :)

**Comment by mdeniz** [comment]
From my side it LGTM beside Henne's nitpicks. Maybe we should care about those c&p lines of code... :smile:

**Comment by hennevogel** [comment]
We're going to iterate over time another time :-)

**Pull request merged**